pythonforandroid/recipes/android/src/android/storage.py

```diff
@@ -0,0 +1,83 @@
+
+from jnius import autoclass, cast
+import os
+
+def app_storage_path():
+    """ Path to internal built-in that is inaccessible/hidden.
+    """
+    from jnius import autoclass, cast
+    PythonActivity = autoclass('org.kivy.android.PythonActivity')
+    currentActivity = cast('android.app.Activity',
+                           PythonActivity.mActivity)
+    context = cast('android.content.ContextWrapper',
+                   currentActivity.getApplicationContext())
+    file_p = cast('java.io.File', context.getFilesDir())
+    return os.path.normpath(os.path.abspath(
+        file_p.getAbsolutePath().replace("/", os.path.sep)))
+
+def primary_external_storage_path():
+    """ Path to built-in SD Card that is externally accessible
+        (through e.g. USB access from a computer),
+        e.g. often /sdcard or /storage/emulated/0
+    """
+    Build = autoclass('android.os.Build')
+    if Build.VERSION.SDK_INT >= 24:
+        Environment = autoclass('android.os.Environment')
+        try:
+            sdpath = (Environment.getExternalStorageDirectory().
+                                  getAbsolutePath())
+            # Apparently this can both return primary (built-in) or
+            # secondary (removable) external storage depending on the device,
+            # therefore check that we got what we wanted:
+            if not Environment.isExternalStorageRemovable(sdpath):
+                return sdpath
+        except Exception:
+            pass
+    if "EXTERNAL_STORAGE" in os.environ:
+        return os.environ["EXTERNAL_STORAGE"]
+    raise RuntimeError(
+        "unexpectedly failed to determine " +
+        "primary external storage path"
+    )
+
+def secondary_external_storage_path():
+    """ Path to removable SD Card which may not be present.
+        e.g. often /sdcard/External_SD/
+    """
+    Build = autoclass('android.os.Build')
+    if Build.VERSION.SDK_INT >= 24:
+        # See if getExternalStorageDirectory() returns secondary ext storage:
+        Environment = autoclass('android.os.Environment')
+        try:
+            sdpath = (Environment.getExternalStorageDirectory().
+                                  getAbsolutePath())
+            # Apparently this can both return primary (built-in) or
+            # secondary (removable) external storage depending on the device,
+            # therefore check that we got what we wanted:
+            if Environment.isExternalStorageRemovable(sdpath):
+                return sdpath
+        except Exception:
```

Comment:
Same here about the Exception.

Reply:
Good catch, also seems addressed, closing